design/boto-design.md

```diff
@@ -0,0 +1,101 @@
+* Isolation - Users should be able to separate access to bucket/prefix between lakefs and s3
+* Users should do the minimal set of changes possible in code to set lakefs
+
+## Non Goals
+* Provide new features that weren't previously possible through lakefs
+
+## Value Propositions
+When a user with an existing python code that uses boto3 client to access s3 wants to move only part of its buckets/prefixes to lakefs, botwo allows using lakefs and s3 together with minimum user's code changes.
+
+## Proposed Design
+The botwo package provides a boto3 like client with two boto3 clients.
+It handles several things for the user:
+* Holds two boto clients and routes the requests between them by the bucket/prefix configuration.
+* Maps 1:1 with the boto AWS service API. All botwo client AWS functions will be generated on the fly according to boto3 client AWS functions.
+* Holds special treatment for two buckets functions like `upload_part_copy`. If more two buckets functions will be added to s3 in the future, we will need to implement them as well.
+* It has more functionality like the event-system that will be implemented by using one of its clients (the default one).
+
+## Alternatives Considered
+
+### Route the requests to lakefs endpoint by the prefix
+```
+s3 = boto3.client('s3')
+if bucket == "bucket_a" and prefix == "my_path":
+  client.endpoint_url  = 'https://lakefs.example.com'
+```
+Does not work because once you initialize boto client with parameters you can’t change them in run time.
+
+### Method Overriding
+
+Monkey patching / overriding base class functions in boto3 with inheritance.
+Does not work because Boto3 functions are generated on the fly from service definitions, so there is no static put_object function to inherit from.
+
+### Creating custom resource-class using the events system
+
+When extending a resource class using the [events system][Extensibility guide], there is no inheritance relationship between the base class and the custom class. It results in a base class enriched with the functions and attributes of the custom class.
+It allows only to replace the bucket and prefix names.
+
+### Route all the requests to lakefs server and implement the bucket configuration there
+It can result in server load because requests like put_object can call directly to s3.
+
+### Wrap boto core fs
+Possible but complex because we will need to implement boto core interface which has a lot of functionality.
+
+## Usage
+
+Current Usage (only with s3):
+```
+import boto3
+
+s3 = boto3.client('s3')
+
+s3.get_object(Bucket="test-bucket", Key="test/object.txt") # in a lot of places
+```
+
+Future Usage: change in every place in the code where there is boto client initialized.
+```
+import boto3
+import botwo
+
+client1 = boto3.client('s3')
+client2 = boto3.client('s3', endpoint_url='https://lakefs.example.com')
+
+mapping = [('bucket-a/*', {'endpoint': 'lakefs', 'bucket': 'my-repo', 'prefix': 'dev/'}),
```

Comment:
About allowing a bucket+prefix combo as the key to routing the request, I'm not sure the benefits justify the complexity.
Also if I understand correctly, we map here from the bucket name to a repo name in lakeFS. This does give some value (no path changes involved) but again makes things more complicated.

Consider the following mapping structure, or something in the middle between this and what you suggested:
```json
[
  {
    "bucket_name": "example-repo",  // will require the user to change a path
    "endpoint": "lakefs",
    "key_prefix": "main/"
  },
  {
    "bucket_name": "example-old-bucket",
    "mapped_bucket_name": "example-repo", // optional - no path change required
    "endpoint": "lakefs",
    "key_prefix": "main/"
  },
  {
    "bucket_name": "*",
    "endpoint": "s3",
    "key_prefix": ""
  }
]

Reply:
Can we perform the routing using just a _function_?  Having to define a complex Pythonic object means it is really hard to understand the object and its logic.  E.g. consider in all of the above: how do I say "all bucket names starting with `ariels-`?  "all JSONs inside bucket `config`?"  **You add syntax.** Now every user suffers from the added syntax, and at the same time users who do need a feature might not have it.  In many cases writing my own function is (much) simpler.

For the cases where you _do_ want to define some structure, we could define **at a second layer** a function `mapping` that takes an object like the above and returns a suitable function.

Alternatively, if we *do* do things like this, it will be easier to understand if we just matched URLs directly, rather than having to split things up like this.  I do not see the advantage of requiring the user to understand "bucket" and "key" separately.

Reply:
All bucket names starting with ariels = ariels*
Can you give an example of such function?